include/freerdp/settings.h

```diff
@@ -74,6 +74,51 @@ extern "C"
 
 	typedef struct rdp_settings rdpSettings;
 
+	/** \brief a BOOL key id and its value
+	 * \since version 3.11.0
+	 */
+	typedef struct
+	{
+		FreeRDP_Settings_Keys_Bool keyId;
+		BOOL value;
+	} FreeRDP_Settings_Bool_Value;
+
+	/** \brief a UINT16 key id and its value
+	 * \since version 3.11.0
+	 */
+	typedef struct
+	{
+		FreeRDP_Settings_Keys_UInt16 keyId;
+		UINT32 value;
+	} FreeRDP_Settings_UInt16_Value;
+
+	/** \brief a INT32 key id and its value
+	 * \since version 3.11.0
+	 */
+	typedef struct
+	{
+		FreeRDP_Settings_Keys_Int32 keyId;
+		UINT32 value;
+	} FreeRDP_Settings_Int32_Value;
+
+	/** \brief a UINT32 key id and its value
+	 * \since version 3.11.0
+	 */
+	typedef struct
+	{
+		FreeRDP_Settings_Keys_UInt32 keyId;
+		UINT32 value;
+	} FreeRDP_Settings_UInt32_Value;
+
+	/** \brief a string key id and its value
+	 * \since version 3.11.0
+	 */
+	typedef struct
+	{
+		FreeRDP_Settings_Keys_String keyId;
+		const char* value;
+	} FreeRDP_Settings_String_Value;
```

Comment:
you can scope all these in a single 
```
struct foo {
int type;
union {
type1 a;
type2 b;
} u;
}
}
```

then you can unify the bulk setters to a single function.

Reply:
Well that can be done as an extension, but I don't think would be that convenient to use: we'll have to specify the type and the field name in the initializer array. The arrays are already massive in `freerdp_settings_new` that would probably double them and have each initialized value spread on multiple lines: probably less readable...

Reply:
@hardening well, not a good argument for the `API` functions, these should be well defined, minimal and complete. (just seen you did not implement all setters here)
by having a single exposed `bulk` setter that could be easily done, including future extensions.

for the initializer stuff in `settings_new`  you can do an internal wrapper though, so you don´t have to add the type.

Reply:
@hardening and just remembered something: check the function `freerdp_settings_set_value_for_name`

Reply:
So we'll have nice code internally thanks to macros and all external users of this function will have long array initializers (or will have to reimplement these macros), that sounds like the bad public `API` recipe.
Another point I've just figured is that with the union we completely loose the static type checking of settings, so why not adding a bulk function for the adventurous, but also keep the functions by types that do the checks at build time...

Reply:
@hardening not if you use named initializers, but yes, without them there are no type checks.
as for external users, where do you need to set these in bulk?

main use is in setting defaults for settings, command line parser and rdp file
and only the default initializers do require bulk initialization.

so, before we add a lot of functions to `API` that we need to maintain I'd want to hear a good reason for that.

Reply:
Well ok, if that kind of thing works, then I'm for it as it does the static check:
```C
typedef struct {
   rdp_settings_type stype;
   union {
       struct {
          FreeRDP_Settings_Keys_UInt16 keyId;
          UINT16 value;
       } t_uint16;
       struct {
          FreeRDP_Settings_Keys_UInt32 keyId;
          UINT32 value;
       } t_uint32;
    ...
   } u;
} FreeRDP_Settings_Value;

#define FREERDP_INIT_UINT32(K, V) { .stype: RDP_SETTINGS_TYPE_UINT32, t_uint32: { FREERDP_##K, V } }

FreeRDP_Settings_Value values[] = {
   FREERDP_INIT_UINT32(FakeMouseMotionInterval, 5),
};


```

Reply:
I´d name the members less similar to the datatypes (e.g. `uint32_t -> u32` and `int32_t -> i32`) but otherwise good.
would also be possible to make the macro a `static inline` function like `static inline FreeRDP_Settings_Value FREERDP_INIT_UINT32(FreeRDP_Settings_Keys_UInt32 key, uint32_t val)`

Reply:
Hum just did a few tests but unfortunately I can see exactly what I was afraid of: the compiler manage to put a `FreeRDP_Settings_Keys_Bool` in a `FreeRDP_Settings_Keys_String` without doing the type check. So for instance 
```
FreeRDP_Settings_Value values[] = {
		FREERDP_SETTING_BOOL(FreeRDP_TcpKeepAlive, TRUE),
		FREERDP_SETTING_STRING(FreeRDP_TcpKeepAlive, FALSE),
	};
```
builds, and without even a warning and so you will figure the error at runtime.

Reply:
Then use a static inline instead of a macro?

Reply:
@hardening also, just tested:
*  `gcc` is a bit less strict than `clang` with warnings, but a integer assignment to a string is always warned about
* assigning a string to an integer value is only warned by `clang -Weverything`

Reply:
Well of course I'm bothered by an int assigned to a string, but I was even more bothered by different enums types assigned to each other without a single notice. That also happens with inline functions. That was my main point against opaque settings: the loss of static checks on the settings, and unfortunately we're at that point :-(